File: src/test/java/masonry/transition/TransitionsTest.java
package masonry.transition;

import masonry.Point;
import net.scriptgate.common.Color3f;
import net.scriptgate.engine.Application;
import net.scriptgate.engine.Engine;
import net.scriptgate.engine.InputComponent;
import net.scriptgate.engine.Renderer;

import java.util.ArrayList;
import java.util.List;

import static masonry.transition.TransitionTrace.trace;
import static net.scriptgate.common.Color3f.RED;
import static net.scriptgate.engine.ApplicationHandlerBuilder.run;
import static net.scriptgate.engine.ApplicationType.OPENGL;

public class TransitionsTest implements Application {

    public static void main(String[] args) {
        run(new TransitionsTest()).in(OPENGL);
    }

    private Point pointFrom;
    private Point pointTo;
    private List<TransitionTrace> traces;


    @Override
    public void initialize() {
        Engine.WIDTH = 256;
        Engine.HEIGHT = 256;

        pointFrom = new Point(16, 16);
        pointTo = new Point(224, 224);
        traces = new ArrayList<>();
        traces.add(trace(new ArcedTransition(pointFrom, pointTo), new Color3f(0, 0, 1)));
        traces.add(trace(new LinearTransition(pointFrom, pointTo), new Color3f(0, 1, 0)));
    }

    @Override
    public void render(Renderer renderer) {
        traces.forEach(trace ->
                trace.render(renderer, point -> renderer.fillCircle(point.x, point.y, 2))
        );

        renderer.setColor(RED);
        renderer.fillCircle(pointFrom.x, pointFrom.y, 2);
        renderer.fillCircle(pointTo.x, pointTo.y, 2);
    }

    @Override
    public void onTick(InputComponent inputComponent, double elapsedTime) {
        traces.forEach(trace -> trace.onTick(elapsedTime));
    }
}

File: src/test/kotlin/net/scriptgate/masonry/transition/TransitionsTest.kt
package net.scriptgate.masonry.transition

import net.scriptgate.common.Color3f
import net.scriptgate.common.Color3f.RED
import net.scriptgate.common.Point
import net.scriptgate.engine.Application
import net.scriptgate.engine.Engine
import net.scriptgate.engine.InputComponent
import net.scriptgate.engine.Renderer
import net.scriptgate.masonry.Launcher.launch
import java.util.*

class TransitionsTest : Application {

    private var pointFrom: Point? = null
    private var pointTo: Point? = null
    private var traces: MutableList<TransitionTrace>? = null


    override fun initialize() {
        Engine.WIDTH = 256
        Engine.HEIGHT = 256

        pointFrom = Point(16, 16)
        pointTo = Point(224, 224)
        traces = ArrayList()
        traces!!.add(TransitionTrace.trace(ArcedTransition(pointFrom!!, pointTo!!), Color3f(0f, 0f, 1f)))
        traces!!.add(TransitionTrace.trace(LinearTransition(pointFrom!!, pointTo!!), Color3f(0f, 1f, 0f)))
    }

    override fun render(renderer: Renderer) {

        traces!!.forEach { trace -> trace.render(renderer) {
            point:Point -> ({ p: Point -> renderer.fillCircle(p.x, p.y, 2)})(point)
        } }

        renderer!!.setColor(RED)
        renderer.fillCircle(pointFrom!!.x, pointFrom!!.y, 2)
        renderer.fillCircle(pointTo!!.x, pointTo!!.y, 2)
    }

    override fun onTick(inputComponent: InputComponent?, elapsedTime: Double) {
        traces!!.forEach { trace -> trace.onTick(elapsedTime) }
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            launch(TransitionsTest())
        }
    }
}
